[Diagnostics] Improve missing conformance diagnostics by using affect…

xedin opened the pull request:
…ed declaration

Instead of simply pointing out which type had conformance failures,
let's use affected declaration instead, which makes diagnostics much
richer e.g.

```
'List<[S], S.Id>' requires that 'S.Id' conform to 'Hashable'
```

versus

```
initializer 'init(_:id:)' requires that 'E' conform to 'Hashable' [with 'E' = 'S.Id']
```

Since latter message uses information about declaration, it can also
point to it in the source. That makes is much easier to understand when
problem is related to overloaded (function) declarations.


<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Diagnostics] Improve missing conformance diagnostics by using affected declaration

Instead of simply pointing out which type had conformance failures,
let's use affected declaration instead, which makes diagnostics much
richer e.g.

```
'List<[S], S.Id>' requires that 'S.Id' conform to 'Hashable'
```

versus

```
initializer 'init(_:id:)' requires that 'E' conform to 'Hashable' [with 'E' = 'S.Id']
```

Since latter message uses information about declaration, it can also
point to it in the source. That makes is much easier to understand when
problem is related to overloaded (function) declarations.

xedin commented:
/cc @slavapestov

xedin commented:
@swift-ci please smoke test

slavapestov commented:
Sweet!

rudkx commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1530,6 +1530,9 @@ ERROR(types_not_equal,none,
       "%0 requires that %1 conform to %2", (Type, Type, Type))
+ERROR(type_does_not_conform_decl_owner,none,
+      "%0 %1 requires that %2 conform to %3 [with %2 = %4]",
+      (DescriptiveDeclKind, DeclName, Type, Type, Type))
```
Reading the output in the tests I find that `struct 'Decl1' requires that 'T' conform to 'P' [with 'T' = 'Y']` is a bit awkward, and I personally don't find referring to the generic type parameter ('T') upfront to be the easiest thing to parse, either.

How about this instead (for `Conforms<Y>.Decl1.self`):
`error: referencing 'Decl1' on 'Conforms<Y>' requires that 'Y' conform to 'P'`
`note: where T = Y` <--- pointing to the `struct Conforms<T>` declaration

?

xedin replied:
I think that's doable if I can detect that Y is an archetype, then I can refer to it's declaration instead, will try...

xedin replied:
Although I think we need to keep `requires that 'T' conform to 'P'` because there is not going to be a `Y` in the declaration error is pointing to.

xedin replied:
or maybe I should just re-do all fo the notes instead of including `[with ...]` into the diagnostic, just add it to the note like you suggested...

rudkx reviewed (approved):
LGTM. As mentioned, I wonder if there might be an even better way to present the information.

xedin committed:
[Diagnostics] Improve missing conformance diagnostics for sub-types and members

If generic parameter associated with missing conformance comes
from different context diagnose the problem as "referencing" a
specific declaration from affected type.

xedin commented:
@rudkx I've updated PR to make it so sub-types diagnosed as "referenced" e.g. 

```
referencing struct 'Decl1' on 'Conforms' requires that 'Y' conform to 'P'
```

or 

```
referencing instance method 'joined()' on 'Collection' requires that 'Int' conform to 'Collection'
```

WDYT?

xedin commented:
@swift-ci please smoke test

rudkx commented:
@xedin Looks great!

xedin commented:
@swift-ci please test
